blockchain: Decouple processing and download logic.

davecgh opened the pull request:
This completely reworks the way block index and processing works to use headers-first semantics and support out of order processing of the associated block data.

This will ultimately provide a wide variety of benefits as it means the entire shape of the block tree can be determined from the headers alone which in turn will allow much more informed decisions to be made, provides better information regarding sync status and warning conditions, and allows future work to add invalidation and reconsideration of arbitrary blocks.

It must be noted that this is a major change to the way the block index and block handling is done and is deeply integrated with the consensus rules, so it will need a significant amount of testing and extremely careful review.  To that end, the changes have been split up into several commits to help ease review, but, unfortunately the vast majority of them really couldn't be reasonably split, so the bulk of hardest to review changes are in a single commit.

Also of note is that there are a lot of assumptions made in the calling code in regards to expected error attribution and the state the chain has at the moment notifications are processed, so all of those semantics have intentionally been retained, even though they might seem a bit out of place now, in order to limit the amount of changes introduced at a time and thus better ensures correctness.

For example, in the future it would probably make more sense to make the notifications entirely asynchronous and for validation failures to be reported via those asynchronous notifications.  However, before that can happen, all callers would first need to be updated to ensure they do not rely on the chain being in the same state it was at the moment the notification was generated.

Another significant change is that the logic that is used to determine if the chain believes it is current is now more robust and takes advantage of the fact the best known header is now available.

In particular, it introduces two new constraints:
- A latching mechanism that prevents the chain from flipping back to being non current unless no new blocks have been seen for an extended period of time
- The chain only latches to current once it is synced to the header with the most cumulative work that is not known to be invalid in addition to the existing conditions

Finally, in addition to all of the existing full block tests, this introduces a comprehensive set of processing order tests which exercise all of the new logic.

### High level overview of the changes:

- Introduce tracking for whether a block is fully linked, meaning it builds on a branch that has block data for all of its ancestors
- Add received order tracking to ensure miners are not able to gain an advantage in terms of chain selection by only advertising a header
- Add several new pieces of information to the block index:
  - Header with the most cumulative work that is not known to be invalid
  - Header with the most cumulative work that is known to be invalid
  - Map of best chain candidates to aid in efficient selection
  - Map of unlinked children for more efficient linking
  - Prunable cached chain tips to significantly reduce potential search space during invalidation
- Introduce a compare function which determines which of two nodes should be considered better for the purposes of best chain selection
- Add chain tip iteration capabilities with potential filtering
- Mark all descendants invalid due to known invalid ancestor when a block is marked invalid
- Add ability to determine if a block can currently be validated
- Rework the chain reorganization func to work with an arbitrary target
- Modify the overall chain reorg func to reorg to the block with the most cumulative work that is valid in the case it is not possible to reorg to the target block
- Add a new `MultiError` type for keeping track of multiple errors in the same call
- Add a new `ErrNoBlockData` error kind
- Update all cases that only require all of the ancestor block data to be available to check for that condition instead of the more strict condition of already being fully validated
- Introduce a new processing lock separate from the overall chain lock
- Add new method and ability to independently accept block headers
- Rework the block processing func to accept blocks out of order so long as their header is already known valid
- Keep a cache of blocks that recently passed contextual validation checks
- Cleanup and correct various comments to match reality
- Rework chain current logic to improve robustness and take advantage new best known header availability
  - Adds a latching mechanism that prevents the chain from flipping back to being non current unless no new blocks have been seen for an extended period of time
  - Only latches the chain to current once it is synced to the header with the most cumulative work that is not known to be invalid in addition to the existing conditions
- Add comprehensive tests to exercise the new processing logic and best header tracking (for both invalid and not known invalid)
  - Support query block test name by hash in the testing
  - Improve test harness logging
  - Add separate test block generation and acceptance
- Update ` README.md`  and `doc.go` to accommodate the changes

---
This is work towards #1145.

davecgh commented:
Oh, one other thing I forgot to call out is that the `NTBlockAccepted` notifications are now delivered at the same time for all connected blocks as opposed to interspersed with every block as it's processed.  This is acceptable because all of the information the code that uses the notification require is available in the notification.

In order to increase confidence in the notifications in general, I also added some logging for them as they're sent and compared the before and after of the full block tests as follows:

```
$ cat testfullblocks.txt | grep NT | cut -d' ' -f2 | sort | uniq -c
    204 NTBlockAccepted
    210 NTBlockConnected
     20 NTBlockDisconnected
     12 NTChainReorgDone
     12 NTChainReorgStarted
    179 NTNewTickets
    226 NTNewTipBlockChecked
      7 NTReorganization
    179 NTSpentAndMissedTickets

$ cat testfullblocksnew.txt | grep NT | cut -d' ' -f2 | sort | uniq -c
    204 NTBlockAccepted
    210 NTBlockConnected
     20 NTBlockDisconnected
     12 NTChainReorgDone
     12 NTChainReorgStarted
    179 NTNewTickets
    226 NTNewTipBlockChecked
      7 NTReorganization
    179 NTSpentAndMissedTickets
```

As can be seen the same number of and type of notifications are being produced.  Closer inspection at the logs shows they are produced in the same order as expected as well.

davecgh commented on blockchain/chain.go:
```diff
@@ -1143,94 +1167,148 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
+	}
+
+	// Determine if there were any reorg errors and either extract and return
+	// the error directly when there was only a singel error or return them all
```
```suggestion
	// the error directly when there was only a single error or return them all
```

davecgh replied:
Updated.

davecgh commented on blockchain/common_test.go:
```diff
@@ -467,10 +470,40 @@ type chaingenHarness struct {
 
+// newChaingenHarnessWithGen creates and returns a new instance of a chaingen
+// harness that encapsulates the provided test instance and existing chaingen
+// generator along with a teardown function  the caller should invoke when done
```
```suggestion
// generator along with a teardown function the caller should invoke when done
```

davecgh replied:
Done.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,324 @@ const (
+}
+
+// maybeAcceptBlockHeader potentially accepts the header to the block index and,
+// if accepted, returns a block node associated with the header.  It performs
```
```suggestion
// if accepted, returns the block node associated with the header.  It performs
```

davecgh replied:
Done.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,324 @@ const (
+}
+
+// ProcessBlockHeader is the main workhorse for handling insertion of new block
+// headers into the block chain using header-first semantics.  It includes
```
```suggestion
// headers into the block chain using headers-first semantics.  It includes
```

davecgh replied:
Updated.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,324 @@ const (
+			return nodes[:i], err
+		}
+
+		// Cache the block and mark is as recently checked to avoid loading and
```
```suggestion
		// Cache the block and mark it as recently checked to avoid loading and
```

davecgh replied:
Done.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +361,219 @@ const (
+	// NOTE: The fast add flag from the caller is used as a standin for now
+	// since that validation is currently handled by the calling code.  In the
+	// future, the calling code should be updated to process all headers through
+	// chain and then the ability to specify the fast add flag should be removed
```
```suggestion
	// this package and then the ability to specify the fast add flag should be removed
```

davecgh replied:
Updated.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +361,219 @@ const (
+	// The final error is a multi error when there is a reorg error and an error
+	// was carried forward from above.  Additionally, in the case the reorg
+	// error is itself a multi error, combine it into a single multi error
+	// rather it inside another one.
```
```suggestion
	// rather than wrapping it inside another one.
```

davecgh replied:
Added.

davecgh reviewed (commented):
Comment typos.

dnldd commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,417 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+
+// AcceptBlockData updates the block index state to account for the full data
+// for a block becoming available.  For example, blocks that are currently not
+// eligible for validation due to either not having having the block data itself
```
s /having having / having

davecgh replied:
Fixed.

dnldd commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,417 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+
+	// The best chain candidates must always contain at least the current best
+	// chain tip.  Assert this assumption is true.
+	if len(bi.bestChainCandidates) == 0 {
```
Perhaps an exception can be made in the work sum comparisons to not remove the current best chain tip if it is ha less work than the passed node to mitigate panicking since the list will be guaranteed to always have one entry?

davecgh replied:
The best chain tip is only know to the chain instance itself as opposed to the block index, so it would have to be passed in.  At that point, you'd end up with the same potential issue of it being incorrect and would ideally want to assert that too, so I don't think it would buy anything.

dnldd commented on blockchain/chain.go:
```diff
@@ -1318,206 +1411,82 @@ func (b *BlockChain) flushBlockIndexWarnOnly() {
-// The flags modify the behavior of this function as follows:
-//  - BFFastAdd: Avoids several expensive transaction validation operations.
-//    This is useful when using checkpoints.
+// isOldTimestamp returns whether the given node has a timestamp is too far in
```
s / ...has a timestamp is too far... / ...has a timestamp too far ...

davecgh replied:
Corrected.

davecgh commented on blockchain/checkpoints.go:
```diff
@@ -93,6 +95,18 @@ func (b *BlockChain) maybeUpdateMostRecentCheckpoint(node *blockNode) {
 }
 
+// isKnownCheckpointAncestor determines whether the provided node is an ancestor
+// of the most recently-known checkpoint.  False is returned wen no checkpoint
```
Updated.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,417 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	if bi.bestHeader.status.KnownInvalid() {
+		// Use the first ancestor of the failed block that is not known to be
+		// invalid as the lower bound for the best header.  This will typically
+		// be parent of the failed block, but it might be some more distant
```
Updated.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,417 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// Start with processing at least the passed node.
+	//
+	// Note that no additional space is preallocated here because it is fairly
+	// rare for there to be more than the single block being linked and thus it
```
Added.

davecgh commented on blockchain/chain.go:
```diff
@@ -1096,19 +1109,19 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
 		}
 
-		// Skip validation if the block is already known to be valid.  However,
+		// Skip validation if the block hash already been validated.  However,
```
Corrected.

sefbkn commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+			bi.setStatusFlags(n, statusInvalidAncestor)
+			bi.removeBestChainCandidate(n)
+
+			// Remove any children that depend on the failed block from set of
```
```suggestion
			// Remove any children that depend on the failed block from the set of
```

davecgh replied:
Added.

sefbkn commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// block data for 102a becoming available and found to be invalid.  Blocks
+	// 106c and 105c would already be marked invalid, and blocks 104c and 103a
+	// need to be marked invalid.
+	markDescendantsFailed := func(node, tip *blockNode) {
```
nit: Since this function flags descendant nodes as `statusInvalidAncestor`, would it be worth calling this `markDescendantsInvalid` rather than `markDescendantsFailed `?

davecgh replied:
Yeah, I agree that is more accurate name.  Updated.

sefbkn commented on blockchain/chainquery.go:
```diff
@@ -132,3 +132,25 @@ func (b *BlockChain) ChainTips() []ChainTipInfo {
+// BestInvalidHeader returns the header with the most cumulative work that is
+// known to be invalid.  It will be a hash of all zeroes if there is no such
+// header.
+func (b *BlockChain) BestInvalidHeader() chainhash.Hash {
```
What is the primary reason for tracking and exposing the best invalid header?

davecgh replied:
For this PR, the main use is in the tests to help ensure all of the logic is working as expected in terms of invalidation.

However, in addition to helping ensure correctness, the main intention is that it will be useful once we transition to solely using headers for network synchronization.  At that point it can be used easily detect when the network has forked away via a hard fork vote and related consensus change which in turn will provide the ability to behave nicely in that circumstances as well as provide more understandable warnings and error messages that they need to upgrade.

sefbkn commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	// is found.  However, that would result in incorrect behavior if a block is
+	// marked invalid deeper in a branch first and then an earlier block is
+	// later marked invalid as it would result in skipping the intermediate
+	// blocks thereby NOT marking them as invalid as they should be.
```
General question around this. This walks backwards from the set of tips to potentially reach the node since there is no direct path to walk from a given node to all descendants. Would maintaining a structure that holds onto both linked and unlinked nodes and gets pruned after a certain distance from the best tip make sense? That way there would not be a need to potentially walk previously invalidated descendants more than once.

davecgh replied:
Perhaps I'm not following your suggestion, but I'm not sure how that would help here.  The highlighted case is not one of being linked or unlinked, it's about the ability for arbitrary nodes along a given path to be invalid without their ancestors being invalid.

It probably isn't super obvious yet because I haven't implemented it yet, but I plan to add the ability to manually invalidate and reconsider arbitrary blocks and I coded this to ensure it will work under under that circumstance.

For example, consider you have some tips:

```
100 -> 101 -> 102 -> 103  -> 104  -> 105  -> 106  -> 107
                 \-> 103a -> 104a -> 105a -> 106a
                         \-> 104b -> 106b
```

Now, you manually mark block 105a invalid.  That will, in turn make 106a invalid due to having an invalid ancestor.  Then, after that, you mark 103a invalid.  Well, if you're trying to skip invalidated descendants, you would skip 106a because it's already invalid, which means you would not mark 104a invalid even though it needs to be.

dnldd commented on blockchain/common_test.go:
```diff
@@ -482,19 +515,94 @@ func newChaingenHarness(t *testing.T, params *chaincfg.Params, dbName string) (*
+	g.t.Logf("Testing accept block header %q (hash %s, height %d)", blockName,
+		blockHash, blockHeight)
+
+	// Determine is the header is already known before attempting to process it.
```
s / Determine is the header is... / Determine the header is...

davecgh replied:
Changed to `Determine if the header ....` and also fixed another case of the same.

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1028 @@ func TestProcessOrder(t *testing.T) {
+
+	// -------------------------------------------------------------------------
+	// Ensure that headers that form a block index with a bunch of branches from
+	// different fork points are accpted and result in the one with the most
```
```suggestion
	// different fork points are accepted and result in the one with the most
```

davecgh replied:
Updated.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -357,6 +403,57 @@ func (node *blockNode) CalcPastMedianTime() time.Time {
+	}
+	if a[index] > b[index] {
+		return 1
+	} else {
+		return -1
+	}
```
Fair point.  Will update.

davecgh replied:
Done.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+		}
+	}
+	bi.cachedTipsStart = height
+	bi.cachedTipsLastPruned = time.Now()
```
I'll update it, but, I actually didn't bother because it really doesn't matter. This method is only ever called when block is connected and the height can only be <= 0 for the first 12 blocks and thus would only be hit during the initial sync and/or the first blocks of simnet anyway.

davecgh replied:
Updated.

rstaudt2 commented on blockchain/common_test.go:
```diff
@@ -515,6 +518,93 @@ func newChaingenHarness(t *testing.T, params *chaincfg.Params, dbName string) (*
+	}
+}
+
+// AcceptBlockData processes the block associated with the given name in the
```
s/AcceptBlockData/AcceptBlockDataWithExpectedTip

davecgh replied:
Whoops.  Nice catch.  Done.

rstaudt2 commented on blockchain/common_test.go:
```diff
@@ -550,6 +640,65 @@ func (g *chaingenHarness) AcceptTipBlock() {
+	}
+
+	// Ensure the rejected header was not added to the block index when it was
+	// not already previously successfully added and that is was not removed if
```
s/is/it

davecgh replied:
Updated.

rstaudt2 reviewed (commented):
Got through the tests so far and called out a few typos.  The process tests read very well and are super helpful for visualizing all of the various scenarios this takes into account!

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1026 @@ func TestProcessOrder(t *testing.T) {
+	// -------------------------------------------------------------------------
+
+	// Accept the block data for b4g, b5g, and b7g but only the block header for
+	// b6h.  The data for block b7g should be accepted because b6g is not yet
```
s/b6h/b6g/

davecgh replied:
Done.

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1026 @@ func TestProcessOrder(t *testing.T) {
+	g.NextBlock("b7i", outs[6], ticketOuts[6])
+	g.NextBlock("b8i", outs[7], ticketOuts[7])
+
+	// Generate a side chain that contains an invalid block nested a several
```
Fixed.

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1026 @@ func TestProcessOrder(t *testing.T) {
+	// complicated branching structure starts.
+	//
+	// All of the headers for these blocks are already known, so this also
+	// exercises in order processing as comapred to the later tests which deal
```
Updated.

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1026 @@ func TestProcessOrder(t *testing.T) {
+	// -------------------------------------------------------------------------
+
+	// Accept the block data for b4i and b5i, but only the block header for b6i.
+	// The header for block b6i should be accepted the block is not yet known to
```
Done.

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1026 @@ func TestProcessOrder(t *testing.T) {
+
+	// Accept the block data for b4j, b5j, and b7j, but only the block header
+	// for b6j.  The block data for b7j should be accepted at this point even
+	// though it is invaid because it can't be checked yet since the data for
```
All set.

davecgh commented on blockchain/blockindex_test.go:
```diff
@@ -310,3 +311,130 @@ func TestAncestorSkipList(t *testing.T) {
 	}
 }
+
+// TestWorkSorterCompare ensure the work soter less and hash comparison
```
Fixed both.

JoeGruffins reviewed (approved):
Awesome.

On testnet there is a lot of output. One warning a few times `[WRN] FEES: Trying to process mined transactions at block 587476 when previous best block was at height 587476` at different heights. Also this info `[INF] CHAN: FORK: Block 0000002b64225f709a777cbb2acdb011705d8f43bbf32888eeed8476c5c4ecf1 (height 587489) forks the chain at height 587488/block 0000009a0b2f93131a56d24b5151337ebcdd259515b0ef9cf204e44469d01636, but does not cause a reorganize` numerous times with different values for the same block. I don't see either of those logs on mainnet, so wondering why such a difference.

All expected?

I also observed a successful reorg happen on testnet.

I am unable to use wallet with the new rpc api version yet, so unable to test sending transactions and such easily.

davecgh commented on blockchain/chain.go:
```diff
@@ -1143,94 +1167,148 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
+			// Send a notification announcing the start of the chain
+			// reorganization.
+			//
+			// Notice that the chain lock is not released before sending the
+			// notification.  This is intentional and must not be changed
+			// without understanding why!
```
The chain lock is held.  Notice the `reorganizeChain` function requires it.  The comment is repeated because it applies to each notification.  Some of them require it to be released and some of them don't.  So, I wanted to be super explicit about which ones do not release it.

Yes, I didn't explain why so that people won't touch it under they understand why things are that way!

davecgh commented on blockchain/common_test.go:
```diff
@@ -723,26 +957,34 @@ func minUint32(a, b uint32) uint32 {
-	if height <= tipHeight {
-		g.t.Fatalf("not possible to advanced to height %d when the current "+
-			"height is already %d", height, tipHeight)
+	// Only allow this to be called with a sane heights.
```
Updated to `sane heights`.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,326 @@ const (
+// checkKnownInvalidBlock returns an appropriate error when the provided block
+// is known to be invalid either due to failing validation itself or due to
+// having a known invalid ancestor (aka being part of an invalid branch).
+func (b *BlockChain) checkKnownInvalidBlock(node *blockNode) error {
```
Done.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,326 @@ const (
+		// practice since a lot of work was expended to create a block that
+		// satisifies the proof of work requirement.
+		//
+		// Notice that the chain lock is not released before sending the
+		// notification.  This is intentional and must not be changed without
+		// understanding why!
```
Deadlocks.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +363,219 @@ const (
+		}
+
+		var forkLen int64
+		if fork := b.bestChain.FindFork(n); fork != nil {
```
Yes.  It is a `chainView` which has its own mutex.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +363,219 @@ const (
+	for _, n := range acceptedNodes {
+		// Skip any blocks which either themselves failed validation or are
+		// descenants of one that failed.
+		if b.index.NodeStatus(n).KnownInvalid() {
```
Yes.  `b.index` is a `blockIndex` which has its own mutex.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +363,219 @@ const (
+	// rather than wrapping it inside another one.
+	case reorgErr != nil && finalErr != nil:
+		var mErr MultiError
+		if errors.As(reorgErr, &mErr) {
+			combined := make([]error, 0, len(mErr)+1)
+			combined = append(combined, finalErr)
+			combined = append(combined, mErr...)
+			finalErr = MultiError(combined)
+		} else {
+			finalErr = MultiError{finalErr, reorgErr}
+		}
```
I originally started to do that, but it's only really needed in this one place and handling it on the `MultiError` itself requires more complicated logic that is less efficient to handle all of the possible cases, so I decided to forego it.

I do agree that if it ends up be needed in multiple places it would make sense to implement a method for it.

davecgh commented:
> On testnet there is a lot of output. One warning a few times `[WRN] FEES: Trying to process mined transactions at block 587476 when previous best block was at height 587476` at different heights. Also this info `[INF] CHAN: FORK: Block 0000002b64225f709a777cbb2acdb011705d8f43bbf32888eeed8476c5c4ecf1 (height 587489) forks the chain at height 587488/block 0000009a0b2f93131a56d24b5151337ebcdd259515b0ef9cf204e44469d01636, but does not cause a reorganize` numerous times with different values for the same block. I don't see either of those logs on mainnet, so wondering why such a difference.
> 
> All expected?

Yes, testnet has a lot of competing blocks show up around the time the difficulty drops.  Mainnet does not have a difficulty reduction like that, so it doesn't happen there.

The fees warning is also normal.  It happens for all forked blocks.  It probably doesn't actually need to be a warning, but the behavior is the same on master.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -438,6 +591,42 @@ func (bi *blockIndex) addNode(node *blockNode) {
+// and best known invalid block as needed.
+//
+// This differs from addNode in that it performs the additional updates to the
+// block index which only apply when nodes are first loaded form the storage.
```
Updated.

davecgh commented:
Rebased.

davecgh commented:
I updated this to modify the `workSorterLess` function to also account for block availability and update the sorting tests accordingly.

The extra condition has no effect when sorting best chain candidates since they are only ever added when their data is available, however, it does matter for determining the best headers (both not known and known to be invalid) with more accuracy.  More specifically, it improves the overall accuracy of determining when the chain is likely to be fully synced and/or when old nodes that have not upgraded are forked from the network.

rstaudt2 reviewed (commented):
This looks great.  I had a few more small comments inline, and outside of that, everything looks good to me.

I also tested a full sync on mainnet and testnet with checkpoints enabled and disabled and tested various scenarios in simnet such as reorgs, and did not see anything unexpected.

davecgh commented on blockchain/chain.go:
```diff
@@ -1143,94 +1167,148 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
 		}
 
-		return reorgErr
+		// Attempt to reorganize to the chain to the new tip.  In the case it
```
Removed.

davecgh commented on blockchain/chain.go:
```diff
@@ -1051,7 +1050,15 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
 
-	// Load the fork block if there are blocks to attach and it's not already
+	// Determine the blocks to attach after the fork point.  Each block is added
+	// to the slice from back to front so that so they are attached in the
```
Fixed.

davecgh commented on blockchain/chain.go:
```diff
@@ -929,66 +940,56 @@ func (b *BlockChain) loadOrCreateFilter(block *dcrutil.Block, view *UtxoViewpoin
 	tip := b.bestChain.Tip()
+	fork := b.bestChain.FindFork(target)
+
+	// Disconnect all of the blocks back to the point of the fork .  This
```
Removed.

davecgh commented on blockchain/error.go:
```diff
@@ -367,6 +376,10 @@ const (
 	// ErrPoolSize indicates an error in the ticket pool size for this block.
 	ErrPoolSize = ErrorKind("ErrPoolSize")
 
+	// ErrForceReorgSameBlock indicates that a reroganization was attempted to
```
Changed.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +365,219 @@ const (
+	// their ancestors available and update the block index state to account for
+	// any that fail validation.
+	//
+	// Note that this done here because it allows any blocks that fail this
```
Added.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +365,219 @@ const (
+	// NOTE: The fast add flag from the caller is used as a standin for now
+	// since that validation is currently handled by the calling code.  In the
+	// future, the calling code should be updated to process all headers through
+	// this pakcage and then the ability to specify the fast add flag should be
```
Fixed

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,328 @@ const (
+		// some of the early ones, but the only effect in that case is
+		// potentially having to load the block and run the context checks again
+		// later.  That said, in practice, eviction of items essentially never
+		// happens under normal operation, especially once the chain is fully is
```
Removed.

davecgh commented on blockchain/process.go:
```diff
@@ -32,13 +35,328 @@ const (
+//
+// It returns those that were accepted along with an error that applies to the
+// first one that failed validation (if any).  This is sufficient because the
+// provided blocks must be all be descendants of previous ones which means all
```
Updated.

davecgh commented on blockchain/process.go:
```diff
@@ -47,48 +365,219 @@ const (
-	return forkLen, nil
+		// This intentionally falls through since the code below must run to
+		// account for any blocks that were accepted even when not all of them
+		// were.
```
No, I don't believe so because it is possible the daemon was interrupted while connecting blocks previously and thus the current tip might actually be behind the latest possible block which the reorganize below will handle to catch up too.

rstaudt2 replied:
Makes sense, thanks for the clarification.  That comment may be slightly misleading then since the code below should run even if _none_ were accepted (by reading only the code I would have had the same conclusion as you but the comment led me to think otherwise).

davecgh replied:
Good point.  I've updated the comment to better call that out.

 		// This intentionally falls through since the code below must run
		// whether or not any blocks were accepted.

davecgh commented on blockchain/blockindex.go:
```diff
@@ -525,6 +733,76 @@ func (bi *blockIndex) removeChainTip(tip *blockNode) {
+// This function MUST be called with the block index lock held (for reads).
+func (bi *blockIndex) forEachChainTipAfterHeight(filter *blockNode, f func(tip *blockNode) error) error {
+	// Use the cached recent chain tips when the filter height permits it.
+	if filter.height >= bi.cachedTipsStart {
```
It is zero by default and all chain tips are added to the cache when initially loading the index from storage and then it is immediately pruned, so, yes, it will always be correct.

Also, great observation on the inclusive starting height!  That is indeed the case.

Updated to use `filter.height >= bi.cachedTipsStart - 1`.

matheusd commented on blockchain/blockindex.go:
```diff
@@ -574,14 +883,424 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+		// be the parent of the failed block, but it might be some more distant
+		// ancestor when performing manual invalidation.
+		n := node.parent
+		for n != nil && n.status.KnownInvalid() {
```
Maybe you also need to include `&& n.status.KnownInvalidAncestor()` here as well? Otherwise performing a manual invalidation might make a header with an invalid parent the best header, unless that is protected at the entrypoint for manual invalidation.

davecgh replied:
`KnownInvalid` includes the invalid ancestor check.

```Go
func (status blockStatus) KnownInvalid() bool {
	return status&(statusValidateFailed|statusInvalidAncestor) != 0
}
```

matheusd replied:
Bleh... :P

matheusd commented on blockchain/blockindex_test.go:
```diff
@@ -310,3 +311,200 @@ func TestAncestorSkipList(t *testing.T) {
+			receivedOrderID: 0,
+		},
+		nodeB: &blockNode{
+			hash:            *higherHash,
+			workSum:         big.NewInt(2),
+			receivedOrderID: 0,
```
`name` doesn't correspond to actual test case (claims nodeA and nodeB have the same data but only nodeA has `statusDataStored`

davecgh replied:
I made the description clearer for this an a few others.  It was meant to "a has same cumulative work", "a has data", "a has same order", etc, but I see that wasn't clear.

matheusd reviewed (commented):
(still reviewing, sending the current items so as to not lose track)

davecgh commented on blockchain/blockindex.go:
```diff
@@ -438,6 +606,42 @@ func (bi *blockIndex) addNode(node *blockNode) {
+	if !node.status.KnownInvalid() && workSorterLess(bi.bestHeader, node) {
+		bi.bestHeader = node
+	}
+
```
Removed.

matheusd reviewed (commented):
(still revewing, running the syncs concurrently)

Got a weird deadlock while running in mainnet. The following dump doesn't show, but the node was stuck for almost an hour, after which I noticed it was stuck and sigaborted it:

https://hastebin.com/jukahudece

There aren't any calls stuck in the blockchain package, so it might not be entirely a problem generated by this PR. One interesting thing to note is that the "Verified checkpoint at height XXX" is duplicated.

matheusd commented on blockchain/process.go:
```diff
@@ -32,13 +35,328 @@ const (
+//
+// It should be noted that this function intentionally does not accept block
+// headers that do not connect to an existing known header or to headers which
+// are already known to part of an invalid branch.  This means headers must be
```
"to **be a** part of an invalid..."

davecgh replied:
Added.

davecgh commented:
> Got a weird deadlock while running in mainnet. The following dump doesn't show, but the node was stuck for almost an hour, after which I noticed it was stuck and sigaborted it:

That is pretty strange looking at the trace, I don't see anything that stands out.  Do you know if it was actually deadlocked or just wasn't continuing to sync?

> 
> ... the "Verified checkpoint at height XXX" is duplicated.

Those have the exact (down to the millisecond) same timestamp.  Did it come out of the log or from a tmux session?  I've noticed sometime tmux can duplicate messages when scrolling.

matheusd commented:
> Do you know if it was actually deadlocked or just wasn't continuing to sync?

Logging simply stopped at that specific line (before the SIGABRT), so I assumed it was deadlocked. I was running with `--debuglevel=info` and didn't think to query peers, current best block, etc, so I can't say with certainty it wasn't just connected to very naughty peers or I if I had a transient network issue.

> Did it come out of the log or from a tmux session?

It's duplicated in the log files as well (checked from outside a tmux session). I consistently get the duplicated msgs on this PR while the current master does not produce them.

matheusd commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1025 @@ func TestProcessOrder(t *testing.T) {
+	g.AcceptHeader("b10")
+	g.AcceptHeader("b10a")
+	g.AcceptHeader("b11")
+	g.ExpectBestHeader("b11")
```
I noticed you're not testing best header selection the same you way you test best invalid header selection and best tip selection.

Do you think it's worth it to add such assertions (for example, by accepting the b#b branch first, then ensuring that accepting the b# branch later correctly selects b11 as best header) or are the tests added for `workSorterLess()` sufficient?

davecgh replied:
I added a bunch of best header tests (both best and best invalid) as well as is current tests in the invalidate/reconsider PR, so I didn't add too many here, but I can go back and add some here too.  It never hurts to further validate assumptions.

matheusd reviewed (commented):
Finished the first pass review. Aside from the minor issues, looks good. Will do another general pass after the duplicate positional checks issue is addressed. I also ran the sync a bunch of times in mainnet.

Just wanna say, this _really_ mustn't have been fun to write, since it involves such a critical piece of the processing code for the blockchain. Definitely was hard to review, since the individual checks are all interdependent in the critical commit.

OTOH, the resulting decoupled processing is a very nice thing to achieve :P

matheusd commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1025 @@ func TestProcessOrder(t *testing.T) {
+//            \-> b4h -> b5h@ -> b6h  -> b7h
+//            \-> b4i -> b5i  -> b6i@ -> b7i  -> b8i
+//            \-> b4j -> b5j  -> b6j  -> b7j@
+func genSharedProcessTestBlocks(t *testing.T) (*chaingen.Generator, error) {
```
I'm guessing you're planning to reuse this in future tests?

davecgh replied:
Correct.  There is another PR up which uses it already (#2536).

matheusd commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1025 @@ func TestProcessOrder(t *testing.T) {
+// variety of fairly complex scenarios selects the expected best chain and
+// properly tracks the header with the most cumulative work that is not known to
+// be invalid as well as the one that is known to be invalid (when it exists).
+func TestProcessLogic(t *testing.T) {
```
Also, just wanna hightlight this is an awesome test :P

davecgh commented on blockchain/process_test.go:
```diff
@@ -138,13 +147,1025 @@ func TestProcessOrder(t *testing.T) {
+//   - b5b is a block with a vote from bad ticket (context failure)
+//   - b6g is a block with a vote from bad ticket (context failure)
+//   - b5h is a block with a double spend (connect failure)
+//   - b6i is a block with a double spend (connect failure)
```
Added.

davecgh commented on blockchain/chain.go:
```diff
@@ -1405,42 +1414,80 @@ func (b *BlockChain) flushBlockIndexWarnOnly() {
+// the current latched state and an additional check which returns false in the
+// case no new blocks have been seen for an extended period of time.
+//
+// The initial factors that are used latch the state to current are:
```
Updated.

davecgh commented on blockchain/treasury.go:
```diff
@@ -519,7 +519,7 @@ type TreasuryBalanceInfo struct {
 func (b *BlockChain) TreasuryBalance(hash *chainhash.Hash) (*TreasuryBalanceInfo, error) {
 	node := b.index.LookupNode(hash)
-	if node == nil || !b.index.NodeStatus(node).HaveData() {
+	if node == nil || !b.index.CanValidate(node) {
```
Right.  Previously, the only way `HaveData` would be true is also if all of the ancestors also had their data and, in most cases, the block had been processed.  However, there was also the case where a side chain block that was never fully validated would pass through the old `HaveData` check and fail below as well.

So, I believe they're actually the same in that regard.  That is to say that blocks that haven't been processed yet will fail in the same way it already used to for side chain blocks that hadn't been processed because they were never connected.

I concur that it isn't really a problem either way because the only time anyone should realistically be querying this is once a block is known to be part of the best chain (e.g. connected).

davecgh commented:
> ... the "Verified checkpoint at height XXX" is duplicated.
> 
> EDIT: I tracked the duplicate print down and it's because the positional block header checks are being run twice. It's not a big deal, but I'll get this updated to avoid doing that.

This has been resolved.  I refactored the checks which apply to the block data to a separate function and call it instead when accepting the block data since the headers have already been checked at that point.

[View Changes](https://github.com/decred/dcrd/compare/600154d30aade9e3c442dfc1580bd9f18431474b..bf07a552f55f432901b0b256f215c28ece359db2).

davecgh commented on blockchain/blockindex.go:
```diff
@@ -574,14 +882,424 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+		// Sort potential best chain candidates in descending order according to
+		// their cumulative proof of work so the highest work candidates are
+		// first.
+		sortedCandidates := make([]*blockNode, 0, numCandidates)
```
Upon reviewing this again, I see that this is a bit wasteful.  There is really no reason to allocate a slice and perform a full sort it here because only a single entry is ever returned.  Instead, it would be more efficient to simply iterate the candidates and find the best one.

Further, this is the only place that the sorting is done, so if it is changed, the `workSorter` implementation to support `sort.Interface` for the `sort.Sort` call would no longer be needed allowing the code to be removed.

davecgh replied:
Done.

matheusd commented:
Did a second pass review of the code (as of now), taking note of the final structure for `ProcessBlock()/ProcessBlockHeader()` for the most important sub-calls and some of the checks performed. Here it is:

```
chain.ProcessBlockHeader()
  * chain.maybeAcceptBlockHeader(checkHeaderSanity == true)
    * checkBlockHeaderSanity()
      - checks PoW
    * chain.checkBlockHeaderPositional()
      - diff retargeting
      - correct height
    * index.AddNode()
      - create or extend header tip
      - maybe change best header
    
chain.ProcessBlock()
  # (block header check stage)
  * checkBlockSanity()
    * checkBlockHeaderSanity()
      - checks PoW
    * checkProofOfStake
    - checks block size < wire.MaxBlockPayload
    * checkTransactionSanity() 					# (regular then stake txs)
    - checks total ticket and revocation # in header
  * chain.maybeAcceptBlockHeader(checkHeaderSanity == false)
    * chain.checkBlockHeaderPositional()
      - checks diff retargeting
      - checks correct height
    * index.AddNode()
      - create or extend header tip
      - maybe change best header
      
  # (block data check stage)
  * chain.maybeAcceptBlockData()    
    * chain.checkBlockDataPositional()
      - checks for expired txs
    * index.AcceptBlockData()
  * chain.maybeAcceptBlocks()
    * chain.checkBlockContext()
      * chain.checkBlockHeaderContext()
        - checks stakediff in header 
        - checks poolsize in header
      * checkTransactionContext()         			# (repeated for every tx and stx)
      - checks stake tx counts limits
      - checks finalized txs
      - checks count sig ops
      - checks tspends only in TVI blocks
      
  # (best chain selection stage)
  * chain.reorganizeChain()
    * chain.reorganizeChainInternal()     			# (repeated until a valid tip is found)
      * utxos.disconnectBlock()					# (when reorging to a different branch)
      * chain.disconnectBlock()
      * chain.checkBlockContext()         			# (if not done yet for some block - see previous call)
      * chain.checkConnectBlock()         			# (called for every block in the new branch)
        * chain.tspendChecks()
          - checks tspend in correct block
          * chain.checkTSpendHasVotes()
          * chain.checkTSPendsExpenditure()
        * utxos.disconnectDisapprovedBlock() 			# (if block votes to disapprove parent block)
        * chain.checkTransactionsAndConnect() 			# (first stake, then regular txs)
        * checkBlockScripts() 					# (first stake, then regular txs)
        * blockOneCoinbasePaysTokens()
      * chain.connectBlock()              			# (called for every block that passes the above)
        * dbPut****()						# (store best state, utxo set, treasury, gcs filter)
        - update in-memory best state
        - send block connected & tickets ntnfs
      * index.RemoveLessWorkCandidates
    * chain.maybeUpdateIsCurrent()
 ```

matheusd reviewed (approved):
Also ran the sync again a few times and executed the test suite on every commit.

Woop!

davecgh commented:
I updated this to ensure the index lock is not held during flushing in case pruned ticket information itself has to be reloaded which itself requires the index lock and to add a couple of trace logging statements for connected and disconnected blocks.

Also, rebased.

sefbkn commented:
This PR looks solid overall. Thank you for taking the time to write this along with clear comments and tests. I am still working through it, but have not seen any issues with the latest code thusfar.

—

On a mostly unrelated note with respect to this PR, while syncing I did notice a log message that appears incorrect. It cropped up while testing this PR so I figured it worth mentioning at least even though it's not coming from this set of changes.

```
[WRN] PEER: Misbehaving peer xxx:9108 (outbound): 0 transactions not found -- ban score increased to 80
...
[INF] SRVR: Banned peer xxx (outbound) for 24h0m0s
```

In server.go -> OnNotFound, the following line
```golang
		reason := fmt.Sprintf("%d %v not found", numBlocks, txStr)
```
seems like it should be
```golang
		reason := fmt.Sprintf("%d %v not found", numTxns, txStr)
```

sefbkn commented on blockchain/chain.go:
```diff
@@ -1143,94 +1170,150 @@ func (b *BlockChain) reorganizeChainInternal(targetTip *blockNode) error {
+	// Potentially update whether or not the chain believes it is current based
+	// on the new tip.  Notice that the tip is reset to whatever the best chain
+	// actually is here versus using the one from above since it might not match
+	// reality if there were errors during reorg.
```
Went over this code a few times, and thanks again for implementing these changes. Only noticed a single comment that read strangely and not worth holding up the PR over.

```suggestion
	// reality if there were errors during the reorg.
```

davecgh replied:
There might have been more than one, so it can't really be "the" reorg.  I think it's correct as it is too, but I went ahead and changed it to "while reorganizing" to avoid any ambiguity.

davecgh commented:
Thanks to everyone for reviewing this.  I know it was not a trivial one to review and took a lot of effort.
